### PR TITLE
Upgrade Jest to 24.0.0-alpha.12

### PR DESCRIPTION
**Pull request by rickhanlonii**
## Overview

Upgrades to the Jest 24 beta release

### CODE REVIEW

File: packages/react-dom/src/__tests__/ReactDOMComponent-test.js [hunk 1]
```diff
@@ -2427,7 +2427,7 @@ describe('ReactDOMComponent', () => {
       const container = document.createElement('div');
 
       ReactDOM.render(<img src={obj} />, container);
-      expect(container.firstChild.src).toBe('hello');
+      expect(container.firstChild.src).toBe('http://localhost/hello');
```
Comment (rickhanlonii) [inline packages/react-dom/src/__tests__/ReactDOMComponent-test.js]: jsdom added localhost as a default hostname

File: packages/react-test-renderer/src/__tests__/ReactTestRendererAsync-test.js [hunk 2]
```diff
@@ -158,7 +158,7 @@ describe('ReactTestRendererAsync', () => {
 
       expect(() =>
         expect(renderer).toFlushAndYieldThrough(['foo', 'baz']),
-      ).toThrow('Expected value to equal:');
+      ).toThrow('Difference:');
```
Comment (rickhanlonii) [inline packages/react-test-renderer/src/__tests__/ReactTestRendererAsync-test.js]: The error matcher string changed, we no longer have the "Expected value to equal" text at the top of error outputs (shoutout to @pedrottimark)

File: scripts/jest/config.base.js [hunk 3]
```diff
@@ -12,7 +12,7 @@ module.exports = {
     '.*': require.resolve('./preprocessor.js'),
   },
   setupFiles: [require.resolve('./setupEnvironment.js')],
-  setupTestFrameworkScriptFile: require.resolve('./setupTests.js'),
+  setupFilesAfterEnv: [require.resolve('./setupTests.js')],
```
Comment (rickhanlonii) [inline scripts/jest/config.base.js]: We changed this setting to allow multiple files and name it something that makes more sense (shoutout to @ddruker)

**Comment by acdlite** [comment]
Before I review this can you explain why we're upgrading to an alpha? I don't have context

**Comment by milesj** [comment]
@acdlite The biggest change in Jest 24 is Babel 7 support, so it would be good to test that it works. Jest 24 official will land soon(tm).

**Comment by acdlite** [comment]
We're not even on Babel 7 yet so I don't get why we're upgrading Jest either. I'm not necessarily opposed, I'm just curious what the motivation is.

**Comment by gaearon** [comment]
I told @rickhanlonii I hit a bug related to Promise rejection that was fixed in master so that's probably why. But agreed we probably should hold off until we use Babel 7. Sorry I didn't realize those changes were related.

**Comment by rickhanlonii** [comment]
@milesj Jest supports babel 7 in Jest 23, in 24 we just _use_ babel 7 internally. There are over [100 features and bug fixes](https://github.com/facebook/jest/blob/44054afb917dd2a02f01adfc292375d5752d77f2/CHANGELOG.md#features) in 24, but I think the biggest is probably handling typescript by default

The context was to get a bug fix to @gaearon, but I'm cool closing until we release Jest 24 (soon). We can use this as a reference for what it would take to upgrade 👍

**Comment by SimenB** [comment]
#12548 must land first - jest 24 won't work with babel 6

(might be possible to hack it by installing `babel-jest@23`, but I doubt you want to go down that rabbit hole)

**Pull request closed without merging**